### PR TITLE
Implement ClassDiscriminatorMode.ALL, .NONE, and .POLYMORPHIC

### PR DESCRIPTION
**Pull request by sandwwraith**

### CODE REVIEW

File: docs/json.md [hunk 1]
```diff
@@ -1296,6 +1336,10 @@ The next chapter covers [Alternative and custom formats (experimental)](formats.
 [JsonBuilder.allowSpecialFloatingPointValues]: https://kotlinlang.org/api/kotlinx.serialization/kotlinx-serialization-json/kotlinx.serialization.json/-json-builder/allow-special-floating-point-values.html
 [JsonBuilder.classDiscriminator]: https://kotlinlang.org/api/kotlinx.serialization/kotlinx-serialization-json/kotlinx.serialization.json/-json-builder/class-discriminator.html
 [JsonClassDiscriminator]: https://kotlinlang.org/api/kotlinx.serialization/kotlinx-serialization-json/kotlinx.serialization.json/-json-class-discriminator/index.html
+[JsonBuilder.classDiscriminatorMode]: https://kotlinlang.org/api/kotlinx.serialization/kotlinx-serialization-json/kotlinx.serialization.json/-json-builder/class-discriminator-mode.html
+[ClassDiscriminatorMode.NONE]: https://kotlinlang.org/api/kotlinx.serialization/kotlinx-serialization-json/kotlinx.serialization.json/-class-discriminator-mode/-n-o-n-e/index.html
+[ClassDiscriminatorMode.POLYMORPHIC]: https://kotlinlang.org/api/kotlinx.serialization/kotlinx-serialization-json/kotlinx.serialization.json/-class-discriminator-mode/-p-o-l-y-m-o-r-p-h-i-c/index.html
```
Comment (shanshin) [inline docs/json.md:1341]: Are it correct links?

Comment (shanshin) [inline docs/json.md:1341]: Are it a correct links?

File: formats/json/commonMain/src/kotlinx/serialization/json/internal/Polymorphic.kt [hunk 2]
```diff
@@ -17,22 +18,40 @@ internal inline fun <T> JsonEncoder.encodePolymorphically(
...
-    checkKind(actualSerializer.descriptor.kind)
-    ifPolymorphic(baseClassDiscriminator)
+    val isPolymorphicSerializer = serializer is AbstractPolymorphicSerializer<*>
+    val needDiscriminator =
+        if (isPolymorphicSerializer) {
+            json.configuration.classDiscriminatorMode != ClassDiscriminatorMode.NONE
+        } else {
+            when (json.configuration.classDiscriminatorMode) {
+                ClassDiscriminatorMode.NONE, ClassDiscriminatorMode.POLYMORPHIC /* already handled in isPolymorphicSerializer */ -> false
+                ClassDiscriminatorMode.ALL_JSON_OBJECTS -> serializer.descriptor.kind in setOf(
```
Comment (shanshin) [inline formats/json/commonMain/src/kotlinx/serialization/json/internal/Polymorphic.kt:32]: Better replace `setOf` by `||` operator. 
I think creating an array and a Set on each JSON object may cause performance degradation.

File: docs/json.md (hunk 1 above)
Comment (sandwwraith) [reply docs/json.md:1341]: They will be correct once we publish the release with this enum. Until then, this file will sit in `dev` and not shown by default

Comment (shanshin) [reply docs/json.md:1341]: ok

**Review by shanshin — APPROVED** [review]

**Pull request merged**